extractor/main.go

```diff
@@ -0,0 +1,293 @@
+				}
+			}
+
+			return nil
+		})
+	}
+
+	return g.Wait()
+}
+
+// writePackageFiles writes all files from a single package to the output directory
+func writePackageFiles(msg vm.MsgAddPackage, outputDir string) error {
+	for _, file := range msg.Package.Files {
+		// Get the output path
+		writePath := filepath.Join(outputDir, file.Name)
+
+		if writeErr := os.WriteFile(writePath, []byte(file.Body), 0644); writeErr != nil {
+			return fmt.Errorf("unable to write file %s, %w", file.Name, writeErr)
+		}
+	}
+
+	return nil
+}
+
+// writePackageMetadata writes the package metadata to the output directory
+func writePackageMetadata(metadata Metadata, outputDir string) error {
+	// Get the output path
+	writePath := filepath.Join(outputDir, packageMetadataFile)
+
+	// Get the JSON metadata
+	metadataRaw, marshalErr := json.Marshal(metadata)
+	if marshalErr != nil {
+		return fmt.Errorf("unable to JSON marshal metadata, %w", marshalErr)
+	}
+
+	if writeErr := os.WriteFile(writePath, metadataRaw, 0644); writeErr != nil {
+		return fmt.Errorf("unable to write package metadata, %w", writeErr)
+	}
+
+	return nil
+}
+
+func extractAddMessages(filePath string) ([]vm.MsgAddPackage, error) {
+	file, openErr := os.Open(filePath)
+	if openErr != nil {
+		return nil, fmt.Errorf("unable to open file, %w", openErr)
+	}
+
+	cleanup := func() error {
+		if closeErr := file.Close(); closeErr != nil {
+			return fmt.Errorf("unable to gracefully close file, %w", closeErr)
+		}
+		return nil
+	}
+
+	reader := bufio.NewReader(file)
+
+	// Used to track what was parsed in the past
+	touchMap := make(map[string]vm.MsgAddPackage)
+
+	// Msg array to be returned for further processing
+	msgArr := make([]vm.MsgAddPackage, 0)
+
+	// Buffer to handle lines longer than 64kb
+	tempBuf := make([]byte, 0)
+
+	for {
+		var tx std.Tx
+		line, isPrefix, err := reader.ReadLine()
+
+		// Exit if no more lines in file
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				break
+			}
+		}
+
+		// If line is too long, save it in a temporary buffer and continue reading line
+		if isPrefix {
+			tempBuf = append(tempBuf, line...)
+			continue
+		}
+
+		// Handle long lines
+		if len(tempBuf) != 0 {
+			// Append last part of line to temporary buffer
+			tempBuf = append(tempBuf, line...)
+
+			// Use line variable to pass it on to amino
+			line = tempBuf
+		}
+
+		if aminoErr := amino.UnmarshalJSON(line, &tx); aminoErr != nil {
+			fmt.Printf("Error while parsing amino JSON at line: %v\nLine:%s\n", aminoErr.Error(), line)
+			continue
+		}
+
+		// Reset tempBuf in case it was used for a long line
+		if tempBuf != nil {
+			tempBuf = nil
+		}
+
+		for _, msg := range tx.Msgs {
+			// Only MsgAddPkg should be parsed
+			if msg.Type() != "add_package" {
+				continue
+			}
+
+			msgAddPkg := msg.(vm.MsgAddPackage)
+
+			path := msgAddPkg.Package.Path
+
+			if _, parsed := touchMap[path]; parsed {
+				// Package already parsed
+				continue
+			}
+
+			touchMap[path] = msgAddPkg
```

Comment:
Following the nit I did before for this map:

```suggestion
			touchMap[path] = true
```